Sources/Sentry/SentrySDK.m

```diff
@@ -134,24 +135,33 @@ + (void)setStartInvocations:(NSUInteger)value
 
 + (void)startWithOptions:(SentryOptions *)options
 {
-    startInvocations++;
+    void (^startSDK)(void) = ^{
+        startInvocations++;
 
-    [SentryLog configure:options.debug diagnosticLevel:options.diagnosticLevel];
+        [SentryLog configure:options.debug diagnosticLevel:options.diagnosticLevel];
 
-    SentryClient *newClient = [[SentryClient alloc] initWithOptions:options];
-    [newClient.fileManager moveAppStateToPreviousAppState];
-    [newClient.fileManager moveBreadcrumbsToPreviousBreadcrumbs];
+        SentryClient *newClient = [[SentryClient alloc] initWithOptions:options];
+        [newClient.fileManager moveAppStateToPreviousAppState];
+        [newClient.fileManager moveBreadcrumbsToPreviousBreadcrumbs];
 
-    SentryScope *scope
-        = options.initialScope([[SentryScope alloc] initWithMaxBreadcrumbs:options.maxBreadcrumbs]);
-    // The Hub needs to be initialized with a client so that closing a session
-    // can happen.
-    [SentrySDK setCurrentHub:[[SentryHub alloc] initWithClient:newClient andScope:scope]];
-    SENTRY_LOG_DEBUG(@"SDK initialized! Version: %@", SentryMeta.versionString);
-    [SentrySDK installIntegrations];
+        SentryScope *scope = options.initialScope(
+            [[SentryScope alloc] initWithMaxBreadcrumbs:options.maxBreadcrumbs]);
+        // The Hub needs to be initialized with a client so that closing a session
+        // can happen.
+        [SentrySDK setCurrentHub:[[SentryHub alloc] initWithClient:newClient andScope:scope]];
+        SENTRY_LOG_DEBUG(@"SDK initialized! Version: %@", SentryMeta.versionString);
+        [SentrySDK installIntegrations];
 
-    [SentryCrashWrapper.sharedInstance startBinaryImageCache];
-    [SentryDependencyContainer.sharedInstance.binaryImageCache start];
+        [SentryCrashWrapper.sharedInstance startBinaryImageCache];
+        [SentryDependencyContainer.sharedInstance.binaryImageCache start];
+    };
+
+    if (NSThread.isMainThread) {
+        startSDK();
+    } else {
+        [SentryDependencyContainer.sharedInstance.dispatchQueueWrapper
+            dispatchAsyncOnMainQueue:startSDK];
```

Comment:
I think I've seen another dispatch invocation (`[SentryThreadWrapper onMainThread:]`) in the other PR 🤔 I guess this one resolving `SentryDependencyContainer.sharedInstance` (which visits some sync lock again) should be dropped completely in favour of wrapping in the `SentryThreadWrapper` type, just so that something so simple as sync/async dispatch to Main thread doesn't have to touch multiple instances around the Sentry scope & be performed with a very clear call path. 👌

Reply:
Thanks for following along here, the more eyes the better 👀 I guess you're talking about this: https://github.com/getsentry/sentry-cocoa/pull/3295/files#diff-1c7b8db05183d8ca8e29ad4450733675d013a6b4065104f150d37d04017a36beR20

That will always immediately return, as it will either directly execute the logic on the main thread if it's called from that context, or will dispatch asynchronously to the main queue otherwise, so we shouldn't see lock contention. The only issue is that if SDK consumers initialize from a non-main context, they can potentially lose information for events that happen very early in their launch sequence, but that should hopefully be a calculated tradeoff for whomever does that. We are going to recommend initializing on the main thread in our docs to avoid that kind of thing. 

We are also working to resolve the lock contention we're seeing in `SentryDependencyContainer` in a third PR: https://github.com/getsentry/sentry-cocoa/pull/3294

Does this all sound correct/acceptable to you?

Reply:
Looking again, just a note for the future: we should have `-[SentryThreadWrapper onMainThread:]` use the dispatch queue wrapper, and then use it here as well since it's the same logic. Doesn't have to be in this PR though, it might involve other code/test changes.